Add qemu cmdline generation framework

BbolroC opened the pull request:
It is just for a preliminary work before a real PR.

pmores committed:
runtime-rs: add storing of default_maxmemory

default_maxmemory hasn't been stored so far in MemoryInfo after parsing
configuration.toml. It will be used by the qemu hypervisor driver though
so make it available.

Signed-off-by: Pavel Mores <pmores@redhat.com>

pmores committed:
runtime-rs: add default rootfs type for qemu

Make sure that rootfs type is known early on even if it's not set in
configuration.toml.

Signed-off-by: Pavel Mores <pmores@redhat.com>

pmores committed:
runtime-rs: use qemu cmdline generation framework to launch VM

Deploy the framework added by the previous commit to generate qemu
command line and launch the VM.

We now properly store the child process object which allows us to
implement remaining Hypervisor functions necessary for a simple but
successful VM lifecycle, get_vmm_master_tid() and stop_vm().

Signed-off-by: Pavel Mores <pmores@redhat.com>

pmores committed:
runtime-rs: add simple Persist implementation for Qemu

This is not necessarily meant to work, just to stub out unimplemented
functionality while focusing on more fundamental things.

Signed-off-by: Pavel Mores <pmores@redhat.com>

pmores committed:
runtime-rs: add simple impls to some of Qemu's Hypervisor functions

The idea of most of these is just to prevent running into todo!()s where
we can at the moment, while implementing the fundamental functionality of
VM launch.

Signed-off-by: Pavel Mores <pmores@redhat.com>

pmores committed:
runtime-rs: add basic implementation of qemu command line generation

This current framework is enough to launch a VM with a simple container
in it (e.g. busybox).

Signed-off-by: Pavel Mores <pmores@redhat.com>

BbolroC committed:
runtime-rs: introduce id to struct VhostVsockPci

This is to define a missing field `id` for the struct VhostVsockPci.
The field is also used to construct qemu parameters for the device.

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

BbolroC committed:
runtime-rs: Separate init_config() from new() for struct VsockDevice

As a follow-up for #8516, guest_cid and vhost_fd are not necessarily instantiated
via new(). Instead, the fields could be initialised later when they are really
used to construct QEMU parameters.
This commit is to separate init_config() from new() to initialise guest_cid
and vhost_fd and leave only the assignment of id for the existing function.

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

BbolroC committed:
runtime-rs: Close vhost_fd after QEMU child runs

vhost_fd is not used anymore once it is passed to a QEMU child process.
This is to close the file descriptor after a qemu command is executed.

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

Apokleos commented on src/libs/kata-types/src/config/default.rs:
```diff
@@ -59,6 +59,7 @@ pub const MAX_DRAGONBALL_VCPUS: u32 = 256;
 pub const MIN_DRAGONBALL_MEMORY_SIZE_MB: u32 = 64;
 // Default configuration for qemu
 pub const DEFAULT_QEMU_BINARY_PATH: &str = "/usr/bin/qemu-system-x86_64";
+pub const DEFAULT_QEMU_ROOTFS_TYPE: &str = "ext4";
```
There'll be two default rootfs types as "ext4", the other one is `DEFAULT_CH_ROOTFS_TYPE`. To simplify this, I suggest merging them into a single public default rootfs type for cases where "ext4" is desired.

BbolroC replied:
yep, I will bring this up to @pmores and change the code. thanks!

Apokleos reviewed (commented):
Thx @BbolroC These three commits are relatively independent. I think they are all good, except for the issue with the rootfs type. But it's not big problem.

BbolroC committed:
runtime-rs: Relocate const VM_ROOTFS_FILESYSTEM_* to kata-types

It is advisable to use a single rootfs type where it is used for a default
file system, rather than to have different const variables referring to the
same value for each hypervisor.

This is to relocate VM_ROOTFS_FILESYSTEM_* const variables to kata-types
and make them used for where a corresponding filesystem type is used.

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

pmores commented:
Hi @BbolroC , I believe the vsock changes should go in as a separate PR since they aren't really tied to qemu in any way.

BbolroC commented:
> Hi @BbolroC , I believe the vsock changes should go in as a separate PR since they aren't really tied to qemu in any way.

Thanks for the comment. Yes, I've raised a separate PR at https://github.com/kata-containers/kata-containers/pull/8672.
